Fix deallocation issue race condition with shape info buffers

agibsonccc opened the pull request:
## What changes were proposed in this pull request?
Fixes a race condition where a shape buffer gets deleted but then used.
This race condition happens when the shape info data buffer is returned
but then deallocated due to the scope ending for the original method.

Also adds a new event logger to track allocations and deallocations.
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [ X] Relevant tests for your changes are passing.

agibsonccc committed:
Add extra case for xw_plus_b
Fix scalar bool eps
Misc test fixes

agibsonccc committed:
Add fixes for paragraphvectors deallocation
Add new event logger for more granular logs including allocations, deallocations

agibsonccc committed:
Add new eventlogger
Event logger contains logging events for allocations, deallocations denominated by object type

agibsonccc committed:
Add new eventlogger
Event logger contains logging events for allocations, deallocations denominated by object type
Add missing cuda deallocation logging

agibsonccc committed:
Add new eventlogger
Event logger contains logging events for allocations, deallocations denominated by object type
Add missing cuda deallocation logging
Set shape info data buffer constant
Add new shouldDeallocate() to deallocatables to control deallocations that should not happen

treo commented on deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/java/org/deeplearning4j/models/paragraphvectors/ParagraphVectors.java:
```diff
@@ -1085,6 +1086,7 @@ public Builder enableScavenger(boolean reallyEnable) {
 
         @Override
         public ParagraphVectors build() {
+            System.out.println(UnifiedProfiler.getInstance().printCurrentStats());
```
Leftover println.

treo commented on nd4j/nd4j-backends/nd4j-backend-impls/nd4j-cuda/src/main/java/org/nd4j/linalg/jcublas/ops/executioner/CudaOpContextDeallocator.java:
```diff
@@ -19,18 +19,36 @@
 
     public CudaOpContextDeallocator(CudaOpContext ctx) {
-        context = (OpaqueContext) ctx.contextPointer();
+        context = (OpaqueContext) ctx.contextPointer(); if(EventLogger.getInstance().isEnabled()) {
```
why put it on the same line?

treo commented on platform-tests/src/test/java/org/deeplearning4j/models/paragraphvectors/ParagraphVectorsTest.java:
```diff
@@ -932,6 +938,79 @@ public void testDirectInference(Nd4jBackend backend) throws Exception {
+    @Tag(TagNames.LARGE_RESOURCES)
+    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
+    @ParameterizedTest
+    public void testParallelLoading(Nd4jBackend backend) throws Exception {
```
The entire test looks like it is filled with remnants of debugging.

agibsonccc committed:
Add new eventlogger
Event logger contains logging events for allocations, deallocations denominated by object type
Add missing cuda deallocation logging
Set shape info data buffer constant
Add new shouldDeallocate() to deallocatables to control deallocations that should not happen

agibsonccc committed:
Clean up nits

treo commented on libnd4j/include/legacy/cpu/NativeOps.cpp:
```diff
@@ -2665,6 +2665,18 @@ void setGraphContextInputBuffer(OpaqueContext *ptr, int index, OpaqueDataBuffer
   if(ptr == nullptr)
     throw std::runtime_error("Context pointer is null!");
+  sd::LongType *shapeInfoCast = reinterpret_cast<sd::LongType *>(shapeInfo);
+  sd_printf("Setting input buffer %d\n",index);
```
printf outside of an error condition.

treo commented on nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/memory/deallocation/DeallocatorService.java:
```diff
@@ -48,7 +48,7 @@ public class DeallocatorService {
         // we need to have at least 2 threads, but for CUDA we'd need at least numDevices threads, due to thread->device affinity
         int numDevices = Nd4j.getAffinityManager().getNumberOfDevices();
-        int numThreads = Math.max(2, numDevices * 2);
+        int numThreads = Math.max(2, numDevices * 2);;
```
double `;`

agibsonccc committed:
Clean up nits
